zend_abstract_interface/symbols/lookup.c

```diff
@@ -0,0 +1,519 @@
+        zai_symbol_updated = false;
+    } zend_end_try();
+
+    if (zai_symbol_updated && !EG(exception)) {
+        zai_sandbox_close(&sandbox);
+        return true;
+    }
+
+    zai_sandbox_close(&sandbox);
+    return false;
+}
+
+static inline void *zai_symbol_lookup_table(HashTable *table, const char *name, size_t length, bool ncase, bool pointer ZAI_TSRMLS_DC) {
+#if PHP_VERSION_ID >= 70000
+    zval *result = zend_hash_str_find(table, name, length);
+#else
+    void *result = NULL;
+
+    zend_hash_find(table, name, length + 1, (void **)&result);
+#endif
+
+    if (!result && ncase) {
+        char *ptr = (char *)pemalloc(length + 1, 1);
+
+        for (uint32_t c = 0; c < length; c++) {
+            ptr[c] = tolower(name[c]);
+        }
+
+        ptr[length] = 0;
+
+#if PHP_VERSION_ID >= 70000
+        result = zend_hash_str_find(table, ptr, length);
+#else
+        zend_hash_find(table, ptr, length + 1, (void **)&result);
+#endif
+
+        pefree(ptr, 1);
+    }
+
+#if PHP_VERSION_ID >= 70000
+    if (pointer && result) {
+        return Z_PTR_P(result);
+    } else {
+        return result;
+    }
+#endif
+    return result;
+}
+
+static inline size_t zai_symbol_lookup_lengthof(zai_string_view *view) {
+    if (!view->len) {
+        return 0;
+    }
+
+    if (memcmp(view->ptr, "\\", sizeof("\\")-1) == SUCCESS) {
+        return view->len - 1;
+    }
+
+    return view->len;
+}
+
+static inline const char* zai_symbol_lookup_startof(zai_string_view *view) {
+    if (!view->len) {
+        return NULL;
+    }
+
+    if (memcmp(view->ptr, "\\", sizeof("\\")-1) == SUCCESS) {
+        return view->ptr + 1;
+    }
+
+    return view->ptr;
+}
+
+static inline zai_string_view* zai_symbol_lookup_key(zai_string_view *namespace, zai_string_view *name, bool lower ZAI_TSRMLS_DC) {
+    zai_string_view *result = pemalloc(sizeof(zai_string_view), 1);
+
+    size_t namespace_length = zai_symbol_lookup_lengthof(namespace);
+    size_t name_length      = zai_symbol_lookup_lengthof(name);
+
+    const char* namespace_start   = zai_symbol_lookup_startof(namespace);
+    const char* name_start        = zai_symbol_lookup_startof(name);
+
+#define CHAR_AT(n) (((char*) result->ptr)[n])
+    if (namespace_length) {
+        result->len = namespace_length + name_length + sizeof("\\")-1;
+        result->ptr = pemalloc(result->len + 1, 2);
+
+        memcpy(&CHAR_AT(0), namespace_start, namespace_length);
+
+        for (uint32_t c = 0; c < namespace_length; c++) {
+            CHAR_AT(c) = tolower(CHAR_AT(c));
+        }
+
+        memcpy(&CHAR_AT(namespace_length), "\\", sizeof("\\")-1);
+        memcpy(&CHAR_AT(namespace_length + (sizeof("\\")-1)), name_start, name_length);
+    } else {
+        result->len = name_length;
+        result->ptr = pemalloc(result->len + 1, 1);
+
+        memcpy(&CHAR_AT(0), name_start, name_length);
+    }
+
+    if (lower) {
+        for (uint32_t c = namespace_length ? namespace_length + (sizeof("\\")-1) : 0;
```

Comment:
redundant condition (the tolower of \\ is also \\)
```suggestion
        for (uint32_t c = namespace_length;
```